coderd/coderdtest/oidctest/idp.go

```diff
@@ -0,0 +1,768 @@
+				return
+			}
+			stateStr, ok := f.codeToStateMap.Load(code)
+			if !assert.True(t, ok, "invalid code") {
+				http.Error(rw, "invalid code", http.StatusBadRequest)
+				return
+			}
+			// Always invalidate the code after it is used.
+			f.codeToStateMap.Delete(code)
+
+			idTokenClaims, ok := f.stateToIDTokenClaims.Load(stateStr)
+			if !ok {
+				t.Errorf("missing id token claims")
+				http.Error(rw, "missing id token claims", http.StatusBadRequest)
+				return
+			}
+			claims = idTokenClaims
+		case "refresh_token":
+			refreshToken := values.Get("refresh_token")
+			if !assert.NotEmpty(t, refreshToken, "refresh_token is empty") {
+				http.Error(rw, "invalid refresh_token", http.StatusBadRequest)
+				return
+			}
+
+			_, ok := f.refreshTokens.Load(refreshToken)
+			if !assert.True(t, ok, "invalid refresh_token") {
+				http.Error(rw, "invalid refresh_token", http.StatusBadRequest)
+				return
+			}
+
+			idTokenClaims, ok := f.refreshIDTokenClaims.Load(refreshToken)
+			if !ok {
+				t.Errorf("missing id token claims in refresh")
+				http.Error(rw, "missing id token claims in refresh", http.StatusBadRequest)
+				return
+			}
+
+			claims = idTokenClaims
+			err := f.hookOnRefresh(getEmail(claims))
+			if err != nil {
+				http.Error(rw, fmt.Sprintf("refresh hook blocked refresh: %s", err.Error()), http.StatusBadRequest)
+				return
+			}
+
+			f.refreshTokensUsed.Store(refreshToken, true)
+			// Always invalidate the refresh token after it is used.
+			f.refreshTokens.Delete(refreshToken)
+		default:
+			t.Errorf("unexpected grant_type %q", values.Get("grant_type"))
+			http.Error(rw, "invalid grant_type", http.StatusBadRequest)
+			return
+		}
+
+		exp := time.Now().Add(time.Minute * 5)
+		claims["exp"] = exp.UnixMilli()
+		email := getEmail(claims)
+		refreshToken := f.newRefreshTokens(email)
+		token := map[string]interface{}{
+			"access_token":  f.newToken(email),
+			"refresh_token": refreshToken,
+			"token_type":    "Bearer",
+			"expires_in":    int64((time.Minute * 5).Seconds()),
+			"id_token":      f.encodeClaims(t, claims),
+		}
+		// Store the claims for the next refresh
+		f.refreshIDTokenClaims.Store(refreshToken, claims)
+
+		rw.Header().Set("Content-Type", "application/json")
+		_ = json.NewEncoder(rw).Encode(token)
+	}))
+
+	mux.Handle(userInfoPath, http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		token, err := f.authenticateBearerTokenRequest(t, r)
+		f.logger.Info(r.Context(), "HTTP Call UserInfo",
+			slog.Error(err),
+		)
+		if err != nil {
+			http.Error(rw, fmt.Sprintf("invalid user info request: %s", err.Error()), http.StatusBadRequest)
+			return
+		}
+
+		email, ok := f.accessTokens.Load(token)
+		if !ok {
+			t.Errorf("access token user for user_info has no email to indicate which user")
+			http.Error(rw, "invalid access token, missing user info", http.StatusBadRequest)
+			return
+		}
+		_ = json.NewEncoder(rw).Encode(f.hookUserInfo(email))
```

Comment:
Follow-up: It would be useful to be able to assert calls to `hookUserInfo`.

Reply:
I can add that :+1:. Would be good if you want to assert `SkipUserInfo` should never hit this.

Reply:
Ah that's fair. 👍